File: ports/nim/src/launch.nim
## launch.nim - sprite-tool launch subcommand
## Creates and configures a sprite with a coding agent, git, beads, and optional checkpointing.

import os, strutils, times, locks
import config, sprite

const launchUsage* = """
Usage: sprite-tool launch [options] <sprite-name> [plan-file]

Options:
  --dry-run              Show what would happen without executing
  --no-checkpoint        Disable auto-checkpointing
  --upload <dir>         Upload a local directory to /home/sprite/<dirname>
                         (repeatable: --upload ./data --upload ./tests)

Environment variables:
  ENV_FILE               Path to .env file (default: ./.env)
  SPRITE_TOKEN           sprites.dev API token (or SPRITES_TOKEN in .env)
  AGENT                  "opencode" (default) or "claude"
  CLAUDE_AUTH            "subscription" (default) or "apikey"
  MODEL                  Model override (see below)
  CHECKPOINT_INTERVAL    Seconds between checkpoints (default: 300 = 5min)

Model examples:
  OpenCode: MODEL=opencode/big-pickle  (free, default)
            MODEL=groq/llama-3.3-70b-versatile  (free w/ GROQ_API_KEY)
            MODEL=openai/gpt-4o
            MODEL=google/gemini-2.5-pro
  Claude:   MODEL=opus, MODEL=sonnet, MODEL=haiku

Examples:
  sprite-tool launch my-project plan.md
  sprite-tool launch --upload ./data my-project plan.md
  sprite-tool launch --upload ./data --upload ./tests my-project plan.md
  sprite-tool launch --dry-run my-project plan.md
"""

# Checkpoint thread state
var
  checkpointLock: Lock
  checkpointRunning: bool = false
  checkpointSprite: string = ""
  checkpointIntervalSecs: int = 300

proc checkpointLoop() {.thread.} =
  ## Background thread that creates periodic checkpoints.
  {.cast(gcsafe).}:
    while true:
      sleep(checkpointIntervalSecs * 1000)
      acquire(checkpointLock)
      let running = checkpointRunning
      let spriteName = checkpointSprite
      release(checkpointLock)
      if not running:
        break
      let timeStr = now().format("HH:mm:ss")
      echo "[checkpoint] Creating checkpoint at " & timeStr & "..."
      let (_, exitCode) = spriteCmdCapture("checkpoint", "create", "-s", spriteName)
      if exitCode == 0:
        echo "[checkpoint] Done."
      else:
        echo "[checkpoint] Failed (non-fatal)."

var checkpointThread: Thread[void]

proc startCheckpointing(spriteName: string, interval: int) =
  initLock(checkpointLock)
  acquire(checkpointLock)
  checkpointRunning = true
  checkpointSprite = spriteName
  checkpointIntervalSecs = interval
  release(checkpointLock)
  createThread(checkpointThread, checkpointLoop)
  echo "Auto-checkpointing every " & $interval & "s"

proc stopCheckpointing() =
  acquire(checkpointLock)
  checkpointRunning = false
  release(checkpointLock)
  joinThread(checkpointThread)
  deinitLock(checkpointLock)

proc runLaunch*(args: seq[string]) =
  ## Parse launch-specific args and execute the launch sequence.
  var
    dryRun = false
    checkpointing = true
    uploadDirs: seq[string] = @[]
    positional: seq[string] = @[]

  var i = 0
  while i < args.len:
    let arg = args[i]
    case arg
    of "--dry-run":
      dryRun = true
    of "--no-checkpoint":
      checkpointing = false
    of "--upload":
      inc i
      if i >= args.len:
        echo "ERROR: --upload requires a directory argument"
        quit(1)
      uploadDirs.add(args[i])
    of "--help", "-h":
      echo launchUsage
      quit(0)
    else:
      if arg.startsWith("--"):
        echo "Unknown option: " & arg
        echo launchUsage
        quit(1)
      positional.add(arg)
    inc i

  if positional.len < 1:
    echo launchUsage
    quit(1)

  let spriteName = positional[0]
  let planFile = if positional.len > 1: positional[1] else: ""

  # Load config from environment + .env
  let cfg = loadConfig()

  # ============================================================
  # 1. Check/install sprite CLI
  # ============================================================
  ensureSpriteCli(dryRun)

  # ============================================================
  # 2. Auth sprite
  # ============================================================
  authSprite(cfg.spriteToken, dryRun)

  # ============================================================
  # 3. Create sprite (or use existing)
  # ============================================================
  if dryRun:
    echo "  [dry-run] Would create or reuse sprite '" & spriteName & "'"
  elif spriteExists(spriteName):
    echo "Sprite '" & spriteName & "' already exists, using it."
  else:
    echo "Creating sprite: " & spriteName
    let rc = spriteCmd("create", "-skip-console", spriteName)
    if rc != 0:
      echo "ERROR: Failed to create sprite"
      quit(1)

  # ============================================================
  # 4. Push .env to sprite
  # ============================================================
  if fileExists(cfg.envFile):
    echo "Pushing " & cfg.envFile & "..."
    discard pushFile(spriteName, cfg.envFile, "/home/sprite/.env", dryRun)

  # ============================================================
  # 5. Push plan file if provided
  # ============================================================
  if planFile != "" and fileExists(planFile):
    echo "Pushing " & planFile & "..."
    discard pushFile(spriteName, planFile, "/home/sprite/plan.md", dryRun)

  # ============================================================
  # 6. Upload directories if provided
  # ============================================================
  for dir in uploadDirs:
    if dirExists(dir):
      let dirName = lastPathPart(dir)
      echo "Uploading directory: " & dir & " -> /home/sprite/" & dirName
      discard pushDir(spriteName, dir, "/home/sprite/" & dirName, dryRun)
    else:
      echo "WARNING: --upload dir '" & dir & "' not found, skipping."

  # ============================================================
  # 7. Setup git + beads
  # ============================================================
  echo "Initializing git..."
  discard sx(spriteName, "cd /home/sprite && git init -b main 2>/dev/null || true", dryRun)

  echo "Installing beads..."
  discard sx(spriteName,
    "curl -fsSL https://raw.githubusercontent.com/steveyegge/beads/main/scripts/install.sh | bash",
    dryRun)

  # ============================================================
  # 8. Install and auth coding agent
  # ============================================================
  if cfg.agent == "claude":
    echo "Setting up claude..."
    discard sx(spriteName,
      "command -v claude >/dev/null 2>&1 || npm install -g @anthropic-ai/claude-code",
      dryRun)

    if cfg.claudeAuth == "subscription":
      let credPath = getHomeDir() / ".claude" / ".credentials.json"
      if fileExists(credPath):
        echo "Copying claude subscription credentials..."
        discard pushFile(spriteName, credPath, "/home/sprite/.claude/.credentials.json", dryRun)
        discard sx(spriteName, "chmod 600 ~/.claude/.credentials.json", dryRun)
      else:
        echo "ERROR: ~/.claude/.credentials.json not found"
        echo "Run 'claude' locally first to authenticate, then re-run this script."
        quit(1)
    elif cfg.claudeAuth == "apikey" and cfg.anthropicApiKey != "":
      echo "Setting ANTHROPIC_API_KEY in sprite..."
      discard sx(spriteName,
        "grep -q ANTHROPIC_API_KEY ~/.bashrc 2>/dev/null || echo 'export ANTHROPIC_API_KEY=\"" &
        cfg.anthropicApiKey & "\"' >> ~/.bashrc",
        dryRun)
    else:
      echo "ERROR: No valid claude auth configured"
      echo "Set CLAUDE_AUTH=subscription (default) or CLAUDE_AUTH=apikey with ANTHROPIC_API_KEY"
      quit(1)

  elif cfg.agent == "opencode":
    echo "Setting up opencode..."
    discard sx(spriteName,
      "[ -x ~/.opencode/bin/opencode ] || curl -fsSL https://opencode.ai/install | bash",
      dryRun)
    discard sx(spriteName,
      "grep -q 'source.*\\.env' ~/.bashrc 2>/dev/null || echo '[ -f /home/sprite/.env ] && set -a && source /home/sprite/.env && set +a' >> ~/.bashrc",
      dryRun)
  else:
    echo "ERROR: Unknown AGENT '" & cfg.agent & "'. Use 'claude' or 'opencode'."
    quit(1)

  # ============================================================
  # 9. Launch agent with plan (or open console)
  # ============================================================
  echo ""
  echo "=========================================="
  echo "Sprite '" & spriteName & "' is ready!"
  var agentLine = "Agent: " & cfg.agent
  if cfg.model != "":
    agentLine &= " (model: " & cfg.model & ")"
  echo agentLine
  if checkpointing:
    echo "Checkpointing: every " & $cfg.checkpointInterval & "s"
  echo "=========================================="

  if dryRun:
    echo ""
    echo "[dry-run] Would launch " & cfg.agent & " with plan. No changes were made."
    quit(0)

  if planFile != "":
    # Start auto-checkpointing before agent runs
    if checkpointing:
      startCheckpointing(spriteName, cfg.checkpointInterval)

    echo "Launching " & cfg.agent & " with plan..."

    if cfg.agent == "claude":
      var modelFlag = ""
      if cfg.model != "":
        modelFlag = "--model " & cfg.model & " "
      discard sx(spriteName,
        "cd /home/sprite && claude " & modelFlag & "-p 'read plan.md and complete the plan please'")

    elif cfg.agent == "opencode":
      let ocModel = if cfg.model != "": cfg.model else: "opencode/big-pickle"
      discard sx(spriteName,
        "set -a && source /home/sprite/.env 2>/dev/null && set +a && cd /home/sprite && ~/.opencode/bin/opencode run -m " &
        ocModel & " 'read plan.md and complete the plan please'")

    # Final checkpoint after agent completes
    if checkpointing:
      stopCheckpointing()
    echo "Creating final checkpoint..."
    let (_, exitCode) = spriteCmdCapture("checkpoint", "create", "-s", spriteName)
    if exitCode == 0:
      echo "Final checkpoint saved."
    else:
      echo "Final checkpoint failed (non-fatal)."
  else:
    echo "Opening console..."
    discard spriteCmd("console", "-s", spriteName)

File: ports/fsharp/Launch.fs
module SpriteTool.Launch

open System
open System.Diagnostics
open System.IO
open System.Threading

let private usage () =
    printfn
        "Usage: sprite-tool launch [options] <sprite-name> [plan-file]\n\
         \n\
         Options:\n\
         \  --dry-run              Show what would happen without executing\n\
         \  --no-checkpoint        Disable auto-checkpointing\n\
         \  --upload <dir>         Upload a local directory to /home/sprite/<dirname>\n\
         \                         (repeatable: --upload ./data --upload ./tests)\n\
         \n\
         Environment variables:\n\
         \  ENV_FILE               Path to .env file (default: ./.env)\n\
         \  SPRITE_TOKEN           sprites.dev API token (or SPRITES_TOKEN in .env)\n\
         \  AGENT                  \"opencode\" (default) or \"claude\"\n\
         \  CLAUDE_AUTH            \"subscription\" (default) or \"apikey\"\n\
         \  MODEL                  Model override (see below)\n\
         \  CHECKPOINT_INTERVAL    Seconds between checkpoints (default: 300 = 5min)\n\
         \n\
         Model examples:\n\
         \  OpenCode: MODEL=opencode/big-pickle  (free, default)\n\
         \            MODEL=groq/llama-3.3-70b-versatile  (free w/ GROQ_API_KEY)\n\
         \            MODEL=openai/gpt-4o\n\
         \            MODEL=google/gemini-2.5-pro\n\
         \  Claude:   MODEL=opus, MODEL=sonnet, MODEL=haiku\n\
         \n\
         Examples:\n\
         \  sprite-tool launch my-project plan.md\n\
         \  sprite-tool launch --upload ./data my-project plan.md\n\
         \  sprite-tool launch --upload ./data --upload ./tests my-project plan.md\n\
         \  sprite-tool launch --dry-run my-project plan.md\n\
         \  MODEL=groq/llama-3.3-70b-versatile sprite-tool launch dev plan.md\n\
         \  AGENT=claude MODEL=sonnet sprite-tool launch dev plan.md"

    exit 1

/// Background checkpoint loop using async computation and CancellationToken.
type CheckpointLoop(spriteName: string, interval: int) =
    let cts = new CancellationTokenSource()

    member _.Start() =
        let token = cts.Token

        let loop =
            async {
                while not token.IsCancellationRequested do
                    do! Async.Sleep(interval * 1000)

                    if not token.IsCancellationRequested then
                        let now = DateTime.Now.ToString("HH:mm:ss")
                        printfn "[checkpoint] Creating checkpoint at %s..." now

                        try
                            let psi = ProcessStartInfo("sprite")
                            psi.UseShellExecute <- false
                            psi.RedirectStandardOutput <- true
                            psi.RedirectStandardError <- true

                            for a in [| "checkpoint"; "create"; "-s"; spriteName |] do
                                psi.ArgumentList.Add a

                            use proc = Process.Start psi
                            proc.WaitForExit()

                            if proc.ExitCode = 0 then
                                printfn "[checkpoint] Done."
                            else
                                printfn "[checkpoint] Failed (non-fatal)."
                        with _ ->
                            printfn "[checkpoint] Failed (non-fatal)."
            }

        Async.Start(loop, token)
        printfn "Auto-checkpointing every %ds (async)" interval

    member _.Stop() =
        if not cts.IsCancellationRequested then
            cts.Cancel()

    interface IDisposable with
        member this.Dispose() = this.Stop()

let run (args: string list) =
    // Load config from .env + environment
    let cfg = Config.load ()

    // Parse flags
    let mutable dryRun = false
    let mutable checkpointing = true
    let mutable uploadDirs: string list = []
    let mutable positional: string list = []
    let mutable i = 0
    let argsArr = args |> List.toArray

    while i < argsArr.Length do
        match argsArr.[i] with
        | "--dry-run" ->
            dryRun <- true
            i <- i + 1
        | "--no-checkpoint" ->
            checkpointing <- false
            i <- i + 1
        | "--upload" ->
            i <- i + 1

            if i >= argsArr.Length then
                eprintfn "Error: --upload requires an argument"
                exit 1

            uploadDirs <- uploadDirs @ [ argsArr.[i] ]
            i <- i + 1
        | "--help"
        | "-h" -> usage ()
        | s when s.StartsWith "--" ->
            eprintfn "Unknown option: %s" s
            usage ()
        | other ->
            positional <- positional @ [ other ]
            i <- i + 1

    if positional.Length < 1 then
        usage ()

    let spriteName = positional.[0]

    let planFile =
        if positional.Length > 1 then
            positional.[1]
        else
            ""

    // Shortcut helpers
    let sx cmd = Sprite.sx spriteName cmd dryRun |> ignore
    let sxPass cmd = Sprite.sxPassthrough spriteName cmd dryRun |> ignore

    let pushFile src dest =
        Sprite.pushFile spriteName src dest dryRun

    let pushDir src dest =
        Sprite.pushDir spriteName src dest dryRun

    // 1. Check/install sprite CLI
    let spriteOnPath =
        try
            let psi = ProcessStartInfo("which")
            psi.UseShellExecute <- false
            psi.RedirectStandardOutput <- true
            psi.RedirectStandardError <- true
            psi.ArgumentList.Add "sprite"
            use proc = Process.Start psi
            proc.WaitForExit()
            proc.ExitCode = 0
        with _ ->
            false

    if not spriteOnPath then
        if dryRun then
            printfn "  [dry-run] Would install sprite CLI"
        else
            printfn "Installing sprite CLI..."

            let psi = ProcessStartInfo("bash")
            psi.UseShellExecute <- false
            psi.ArgumentList.Add "-c"
            psi.ArgumentList.Add "curl -fsSL https://sprites.dev/install.sh | sh"
            use proc = Process.Start psi
            proc.WaitForExit()

            let localBin =
                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".local", "bin")

            let currentPath = Environment.GetEnvironmentVariable "PATH"
            Environment.SetEnvironmentVariable("PATH", sprintf "%s:%s" localBin currentPath)

    // 2. Auth sprite
    if not (String.IsNullOrEmpty cfg.SpriteToken) then
        printfn "Authenticating sprite with token..."

        if not dryRun then
            let psi = ProcessStartInfo("sprite")
            psi.UseShellExecute <- false

            for a in [| "auth"; "setup"; "--token"; cfg.SpriteToken |] do
                psi.ArgumentList.Add a

            use proc = Process.Start psi
            proc.WaitForExit()
    else
        printfn "No SPRITE_TOKEN set. Running interactive login..."

        if not dryRun then
            let psi = ProcessStartInfo("sprite")
            psi.UseShellExecute <- false
            psi.ArgumentList.Add "login"
            use proc = Process.Start psi
            proc.WaitForExit()

    // 3. Create sprite (or use existing)
    if dryRun then
        printfn "  [dry-run] Would create or reuse sprite '%s'" spriteName
    elif Sprite.spriteExists spriteName then
        printfn "Sprite '%s' already exists, using it." spriteName
    else
        printfn "Creating sprite: %s" spriteName

        let psi = ProcessStartInfo("sprite")
        psi.UseShellExecute <- false

        for a in [| "create"; "-skip-console"; spriteName |] do
            psi.ArgumentList.Add a

        use proc = Process.Start psi
        proc.WaitForExit()

    // 4. Push .env to sprite
    if File.Exists cfg.EnvFile then
        printfn "Pushing %s..." cfg.EnvFile
        pushFile cfg.EnvFile "/home/sprite/.env"

    // 5. Push plan file if provided
    if not (String.IsNullOrEmpty planFile) && File.Exists planFile then
        printfn "Pushing %s..." planFile
        pushFile planFile "/home/sprite/plan.md"

    // 6. Upload directories if provided
    uploadDirs
    |> List.iter (fun dir ->
        if Directory.Exists dir then
            let dirName = Path.GetFileName(Path.GetFullPath dir)
            printfn "Uploading directory: %s -> /home/sprite/%s" dir dirName
            pushDir dir (sprintf "/home/sprite/%s" dirName)
        else
            printfn "WARNING: --upload dir '%s' not found, skipping." dir)

    // 7. Setup git + beads
    printfn "Initializing git..."
    sx "cd /home/sprite && git init -b main 2>/dev/null || true"

    printfn "Installing beads..."
    sx "curl -fsSL https://raw.githubusercontent.com/steveyegge/beads/main/scripts/install.sh | bash"

    // 8. Install and auth coding agent
    if cfg.Agent = "claude" then
        printfn "Setting up claude..."
        sx "command -v claude >/dev/null 2>&1 || npm install -g @anthropic-ai/claude-code"

        if cfg.ClaudeAuth = "subscription" then
            let credsPath =
                Path.Combine(
                    Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                    ".claude",
                    ".credentials.json"
                )

            if File.Exists credsPath then
                printfn "Copying claude subscription credentials..."
                pushFile credsPath "/home/sprite/.claude/.credentials.json"
                sx "chmod 600 ~/.claude/.credentials.json"
            else
                eprintfn "ERROR: ~/.claude/.credentials.json not found"
                eprintfn "Run 'claude' locally first to authenticate, then re-run this script."
                exit 1
        elif cfg.ClaudeAuth = "apikey" && not (String.IsNullOrEmpty cfg.AnthropicApiKey) then
            printfn "Setting ANTHROPIC_API_KEY in sprite..."

            sx
                (sprintf
                    "grep -q ANTHROPIC_API_KEY ~/.bashrc 2>/dev/null || echo 'export ANTHROPIC_API_KEY=\"%s\"' >> ~/.bashrc"
                    cfg.AnthropicApiKey)
        else
            eprintfn "ERROR: No valid claude auth configured"
            eprintfn "Set CLAUDE_AUTH=subscription (default) or CLAUDE_AUTH=apikey with ANTHROPIC_API_KEY"
            exit 1
    elif cfg.Agent = "opencode" then
        printfn "Setting up opencode..."
        sx "[ -x ~/.opencode/bin/opencode ] || curl -fsSL https://opencode.ai/install | bash"

        sx
            "grep -q 'source.*\\.env' ~/.bashrc 2>/dev/null || echo '[ -f /home/sprite/.env ] && set -a && source /home/sprite/.env && set +a' >> ~/.bashrc"
    else
        eprintfn "ERROR: Unknown AGENT '%s'. Use 'claude' or 'opencode'." cfg.Agent
        exit 1

    // 9. Launch agent with plan (or open console)
    printfn ""
    printfn "=========================================="
    printfn "Sprite '%s' is ready!" spriteName

    let modelNote =
        if String.IsNullOrEmpty cfg.Model then
            ""
        else
            sprintf " (model: %s)" cfg.Model

    printfn "Agent: %s%s" cfg.Agent modelNote

    if checkpointing then
        printfn "Checkpointing: every %ds" cfg.CheckpointInterval

    printfn "=========================================="

    if dryRun then
        printfn ""
        printfn "[dry-run] Would launch %s with plan. No changes were made." cfg.Agent
    elif not (String.IsNullOrEmpty planFile) then
        // Start auto-checkpointing before agent runs
        let mutable checkpointLoop: CheckpointLoop option = None

        if checkpointing then
            let loop = new CheckpointLoop(spriteName, cfg.CheckpointInterval)
            loop.Start()
            checkpointLoop <- Some loop

        printfn "Launching %s with plan..." cfg.Agent

        if cfg.Agent = "claude" then
            let modelFlag =
                if String.IsNullOrEmpty cfg.Model then
                    ""
                else
                    sprintf "--model %s " cfg.Model

            sxPass (sprintf "cd /home/sprite && claude %s-p 'read plan.md and complete the plan please'" modelFlag)
        elif cfg.Agent = "opencode" then
            let ocModel =
                if String.IsNullOrEmpty cfg.Model then
                    "opencode/big-pickle"
                else
                    cfg.Model

            sxPass
                (sprintf
                    "set -a && source /home/sprite/.env 2>/dev/null && set +a && cd /home/sprite && ~/.opencode/bin/opencode run -m %s 'read plan.md and complete the plan please'"
                    ocModel)

        // Final checkpoint after agent completes
        checkpointLoop |> Option.iter (fun loop -> loop.Stop())

        printfn "Creating final checkpoint..."

        let psi = ProcessStartInfo("sprite")
        psi.UseShellExecute <- false
        psi.RedirectStandardOutput <- true
        psi.RedirectStandardError <- true

        for a in [| "checkpoint"; "create"; "-s"; spriteName |] do
            psi.ArgumentList.Add a

        use proc = Process.Start psi
        proc.WaitForExit()

        if proc.ExitCode = 0 then
            printfn "Final checkpoint saved."
        else
            printfn "Final checkpoint failed (non-fatal)."
    else
        printfn "Opening console..."

        let psi = ProcessStartInfo("sprite")
        psi.UseShellExecute <- false

        for a in [| "console"; "-s"; spriteName |] do
            psi.ArgumentList.Add a

        use proc = Process.Start psi
        proc.WaitForExit()
